File: benchmark_release/public/case_000060/src/app/services/auditService.js
// Audit trail service — writes structured events to audit_log table
// NOTE: high-volume environments should batch writes; see PORTAL-198 for async queue design

/**
 * Record a single audit event.
 * @param {object} dbModule - module export from db.js
 * @param {object} event - { actor, action, target }
 */
function record(dbModule, { actor, action, target }) {
  try {
    const conn = dbModule.db || dbModule;
    conn.prepare(
      'INSERT INTO audit_log (actor, action, target) VALUES (?, ?, ?)'
    ).run(actor || 'anonymous', action, target || '');
  } catch (e) {
    // non-fatal — log to stderr and continue; do NOT surface to caller
    console.error('[audit] write failed:', e.message);
  }
}

/**
 * Retrieve recent audit entries, newest first.
 * @param {object} dbModule
 * @param {number} limit - max rows to return (capped at 500)
 */
function recent(dbModule, limit) {
  const cap = Math.min(limit || 50, 500);
  const conn = dbModule.db || dbModule;
  return conn.prepare(
    'SELECT id, actor, action, target, ts FROM audit_log ORDER BY ts DESC LIMIT ?'
  ).all(cap);
}

module.exports = { record, recent };

File: benchmark_controls_release/public/case_000007_clean/src/app/services/audit_service.py
import logging
from flask import request

logger = logging.getLogger("toolforge.audit")


def record(conn, actor: str, action: str, target: str, detail: str = ""):
    ip = request.remote_addr if request else "unknown"
    try:
        conn.execute(
            "INSERT INTO audit_log(actor,action,target,detail,ip_addr) VALUES(?,?,?,?,?)",
            (actor, action, target, detail, ip)
        )
    except Exception as exc:
        logger.warning("audit write failed: %s", exc)